### ISSUE TITLE
How does yt-dlp handle tracking links like `?utm=` and `?si=` ? Is user privacy preserved?

### ISSUE THREAD

**Issue by barkoder** [yt-dlp/yt-dlp#9815]
### DO NOT REMOVE OR SKIP THE ISSUE TEMPLATE

- [X] I understand that I will be **blocked** if I *intentionally* remove or skip any mandatory\* field

### Checklist

- [X] I'm asking a question and **not** reporting a bug or requesting a feature
- [X] I've looked through the [README](https://github.com/yt-dlp/yt-dlp#readme)
- [X] I've verified that I have **updated yt-dlp to nightly or master** ([update instructions](https://github.com/yt-dlp/yt-dlp#update-channels))
- [X] I've searched [known issues](https://github.com/yt-dlp/yt-dlp/issues/3766) and the [bugtracker](https://github.com/yt-dlp/yt-dlp/issues?q=) for similar questions **including closed ones**. DO NOT post duplicates
- [X] I've read the [guidelines for opening an issue](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#opening-an-issue)

### Please make sure the question is worded well enough to be understood

Does it handle those links as is, or does yt-dlp clean the URLs off any tracking parameters before parsing them?

I really don't want youtube tracking me with their `?si` URLs and other websites tracking me with their `?utm` URLs.

Thanks.

### Provide verbose output that clearly demonstrates the problem

- [X] Run **your** yt-dlp command with **-vU** flag added (`yt-dlp -vU <your command line>`)
- [ ] If using API, add `'verbose': True` to `YoutubeDL` params instead
- [X] Copy the WHOLE output (starting with `[debug] Command-line config`) and insert it below

### Complete Verbose Output

```shell
I don't think verbose output is required here but I'm using yt-dlp_min.exe version 2024.04.28.232723
```

**Comment by bashonly** [comment]
It depends on the site that the URL belongs to, which extractor yt-dlp uses for that site, and how that extractor is written.

For Youtube, yt-dlp extracts the video ID from the URL and constructs a clean URL with which it requests the webpage:
https://github.com/yt-dlp/yt-dlp/blob/ac817bc83efd939dca3e40c4b527d0ccfc77172b/yt_dlp/extractor/youtube.py#L4072-L4079
...so all tracking params are stripped.

Some extractors, such as the Reddit extractor, don't request the webpage at all, they just extract the necessary information from the URL and use it to construct an API call:
https://github.com/yt-dlp/yt-dlp/blob/ac817bc83efd939dca3e40c4b527d0ccfc77172b/yt_dlp/extractor/reddit.py#L200-L204
...so that's all good.

But other extractors may pass the input URL to the request verbatim; e.g. this is what the generic extractor does, since it doesn't know which query params are necessary vs meaningless tracking.

Bottom line is, if you care about privacy, then choose to do one of the following:
- strip the tracking params yourself
- check the extractor code for the site in question to see if you can get away without manually stripping them yourself
- if an extractor can strip tracking params but currently doesn't, then submit a PR to improve said extractor

**Comment by barkoder** [comment]
Thank you @bashonly for that comprehensive answer.

But is there a (non-dump-pages) way to know the final URL(after any/all cleaning) before yt-dlp connects to the internet?

If I specify `-v --dump-pages` , I can see that yt-dlp is properly connecting to the clean json URL of reddit, but using `--dump-pages` is very verbose.

    [Reddit] Extracting URL: https://old.reddit.com/r/Damnthatsinteresting/comments/rn4090?utm_source=reddit.com
    [Reddit] rn4090: Downloading JSON metadata
    director: Handler preferences for this request: urllib=0, requests=100, websockets=0
    director: Checking if "requests" supports this request.
    director: Sending request via "requests"
    requests: Starting new HTTPS connection (1): old.reddit.com:443
    [Reddit] Dumping request to https://old.reddit.com/r/Damnthatsinteresting/comments/rn4090/.json
    send: b'GET /r/Damnthatsinteresting/comments/rn4090/.json HTTP/1.1\r\nHost: old.reddit.com\r\nConnection: keep-alive\r\nUser-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/94.0.4606.61 Safari/537.36\r\nAccept: text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8\r\nAccept-Language: en-us,en;q=0.5\r\nSec-Fetch-Mode: navigate\r\nAccept-Encoding: gzip, deflate, br\r\n\r\n

I would appreciate it, if I merely specify `-v`(without the very verbose `--dump-pages`) flag, the log said something like `[Reddit] rn4090: Downloading JSON metadata from \<CLEAN_REDDIT_URL/.json\>` . Instead of requiring `--dump-pages` (super verbose) to get that URL.(Unless there's another way to get that URL.)

Basically, if a given extractor has URL cleaning built-in, and cleans the URL, specifying `-v` should echo the cleaned URL.

Like `Extracting Clean URL: https://example.com/cleanurl`

I think it would help if this was so for all extractors.

The reason I had this question to begin with, was because the `Extracting URL` part of the log contains the URL *with* the tracking params, and I was concerned.

Thanks.

**Comment by pukkandan** [comment]
`--print-traffic`

**Comment by barkoder** [comment]
I should've mentioned that I *had* actually used `--print-traffic` in the previous comment as you can probably tell from the `send: b'GET /r/Damnthatsinteresting/comments/` which isn't part of `--dump-pages`. In the previous comment, I used `--print-traffic` along with `-v --dump-pages` , and the `Dumping request to` line showed up first in the log , so that's what I thought I should highlight.

Anyway, that works so thanks @pukkandan, but `--print-traffic` even without `-v --dump-pages` is still quite verbose. Is it possible to just have the clean URL for that `send` request show up in `-v` along with `https://www.reddit.com` as prefix?

Thanks.

Also is there a way to use `-s` but not have yt-dlp connect to the internet? An offline version of `-s` . I just want to know where yt-dlp is gonna go after it's composed the request, without it actually going to the link.

Thanks.

**Comment by dirkf** [comment]
The various download methods called by extractors have a parameter `note` to describe what is being downloaded that is displayed using `to_screen()`. It wouldn't be a stretch  under `--verbose` to add the actual URL used, or else to have s/t like `---verbose-url`.

The code in `_request_webpage()` that handles this, both here and upstream, is concerned that `note` might be something other than `None` or a `str` (resp. `compat_str`). This seems unnecessary today.

**Comment by bashonly** [comment]
>  It wouldn't be a stretch  under `--verbose` to add the actual URL used, or else to have s/t like `---verbose-url`.
> 
> The code in `_request_webpage()` that handles this

What about redirects? It would be misleading. A simplified and inaccurate `--print-traffic` isn't very useful IMO

**Comment by dirkf** [comment]
The program can report what it's sending, which is what OP is worried about, rather than where the resource actually accessed was. Regarding the latter, extractor code should often check for redirection as a sign of expired or otherwise invalid content. You might consider that `POST` data could be reported too, but that would typically be encoded by the caller and so less transparent.

**Comment by barkoder** [comment]
> or else to have s/t like `--verbose-url`.

I don't know what s/t is(*I tried looking it up and I couldn't find it so do let me know*) but yes please. I'd much prefer a separate option. This way I can lower verbosity even further should I not need `-v` sometimes.

Also let me know about an offline version of `-s --verbose-url` that prints said `--verbose-url` (in reddit's case presumably the json url) and exits with no packets sent anywhere.

Thanks.

**Comment by dirkf** [comment]
s/t = something

Trial upstream implementation:
```console
$ python -m youtube_dl -vF 'BaW_jenozKc'
[debug] System config: []
[debug] User config: []
[debug] Custom config: []
[debug] Command-line args: [u'-vF', u'BaW_jenozKc']
[debug] Encodings: locale UTF-8, fs UTF-8, out UTF-8, pref UTF-8
[debug] youtube-dl version 2021.12.17
[debug] Git HEAD: ac469f627
[debug] Python 2.7.15 (CPython i686 32bit) - Linux-6.1.0-20-686-pae-i686-with-debian-12.5 - OpenSSL 1.1.1a  20 Nov 2018 - glibc 2.1.3
[debug] exe versions: ffmpeg 5.1.4-0, ffprobe 5.1.4-0
[debug] Proxy map: {}
[youtube] BaW_jenozKc: Downloading webpage:
<https://www.youtube.com/watch?v=BaW_jenozKc&bpctr=9999999999&has_verified=1>
[youtube] BaW_jenozKc: Downloading player 8fc6998a:
<https://www.youtube.com/s/player/8fc6998a/player_ias.vflset/en_US/base.js>
[debug] [youtube] Decrypted nsig 7DwIic66gJDGXJJc => 4yM5T1TMEb6frA
[debug] [youtube] Decrypted nsig 33So2BVMwzZiScGw => GFDuV-efKD2y9g
[info] Available formats for BaW_jenozKc:
format code  extension  resolution note
249          webm       audio only audio_quality_low   48k , webm_dash container, opus  (48000Hz), 57.97KiB
...
$
```

**Comment by pukkandan** [comment]
If the objective is to know whether private data is being sent, I don't see how this's useful. The data could be in headers, POST data, etc

> not have yt-dlp connect to the internet

Block in firewall? Or pass an invalid proxy. Each extraction typically has multiple network requests, each of which cannot be performed without the previous passing. So this will only show the first request, which is pointless imo
